[WIP] V4 Manifest Read Support

anoopj opened the pull request:
WIP PR for [s.apache.org/iceberg-single-file-commit](https://s.apache.org/iceberg-single-file-commit)

Implemented so far:

* Foundational types such as TrackedFile interface as unified representation for all V4 entry types
* Reader and basic root manifest expansion

anoopj committed:
API: Add V4 manifest TrackedFile interfaces and implementation

 Introduces the foundational types for V4 manifest format support:
  - TrackedFile interface as unified representation for all V4 entry types
  - DeletionVector and ManifestStats interfaces
  - GenericTrackedFile implementation and test

anoopj committed:
Fix checkstyle error

anoopj committed:
Add position, setters

anoopj committed:
Fix checkstyle error

anoopj committed:
implement reader

anoopj committed:
Checkstyle and formatter

anoopj committed:
Rename file readers

anoopj committed:
implement root manifest reader

anoopj committed:
implement adapter for asData and asDelete

anoopj committed:
Trigger PR refresh

anoopj committed:
Implement manifest expander

anoopj committed:
Support for inline metadata DVs: 32 bit roaring bitmaps for now

anoopj committed:
Stub for snapshot integration

anoopj committed:
Implement parallel expansion

rdblue commented on api/src/main/java/org/apache/iceberg/FileContent.java:
```diff
@@ -18,11 +18,27 @@
+   * Manifest deletion vector entry (V4+ only) - marks entries in a manifest as deleted without
+   * rewriting the manifest.
+   */
+  MANIFEST_DV(5);
```
I prefer the option of having the DV located in a field of the data or delete manifest record. That way we don't have to wait to find the DV before processing a manifest file. Not sure what others think here, but since the DV metadata/content is likely going to be different between the Metadata DV (inline) and Data DV (stored in Puffin), I don't see much value in trying to reuse metadata fields for it.

rdblue replied:
I just check with Amogh and I don't think that this has been firmly decided yet. From the implementation here, I think we should embed the DV in manifest-specific metadata.

rdblue commented on api/src/main/java/org/apache/iceberg/DeletionVector.java:
```diff
@@ -0,0 +1,65 @@
+   * <p>When present, the deletion vector is stored inline in the manifest rather than in a separate
+   * Puffin file.
+   */
+  ByteBuffer inlineContent();
```
I mentioned this in my comment below, but I don't think there's much value in combining the inline MDV metadata and fields to track data DVs stored in Puffin. These aren't overlapping, so I'd keep them separate.

rdblue commented on api/src/main/java/org/apache/iceberg/TrackedFile.java:
```diff
@@ -0,0 +1,268 @@
+          104,
+          "file_size_in_bytes",
+          Types.LongType.get(),
+          "Total file size in bytes. Must be defined if location is defined");
```
Same here.

anoopj replied:
Done

rdblue commented on api/src/main/java/org/apache/iceberg/TrackedFile.java:
```diff
@@ -0,0 +1,268 @@
+   * <p>Contains status, snapshot ID, sequence numbers, and first-row-id. Optional - may be null if
+   * tracking info is inherited.
+   */
+  TrackingInfo trackingInfo();
```
I think that this also requires a field to be defined, so we have a record of the ID used for it.

pvary replied:
What is `content_slice`?

rdblue commented on core/src/main/java/org/apache/iceberg/DataTableScan.java:
```diff
@@ -63,8 +63,17 @@ protected TableScan newRefinedScan(Table table, Schema schema, TableScanContext
   @Override
   public CloseableIterable<FileScanTask> doPlanFiles() {
     Snapshot snapshot = snapshot();
-
```
Nit: Avoid unnecessary whitespace changes. They cause conflicts.

anoopj commented on api/src/main/java/org/apache/iceberg/TrackedFile.java:
```diff
@@ -0,0 +1,268 @@
+          100,
+          "location",
+          Types.StringType.get(),
+          "Location of the file. Optional if content_type is 5 and deletion_vector.inline_content is not null");
```
Yes.

anoopj committed:
implement copyWithStats with IDs

rdblue commented on api/src/main/java/org/apache/iceberg/TrackedFile.java:
```diff
@@ -0,0 +1,268 @@
+   * @throws IllegalStateException if content_type is not DATA
+   * @throws UnsupportedOperationException if ContentStats not yet implemented
+   */
+  DataFile asDataFile(PartitionSpec spec);
```
For this API, I think it is cleaner to call `asDataFile()` without any arguments.

rdblue commented on api/src/main/java/org/apache/iceberg/TrackedFile.java:
```diff
@@ -0,0 +1,268 @@
+  DeleteFile asDeleteFile(PartitionSpec spec);
+
+  /** Set the status for this tracked file entry. */
+  void setStatus(TrackingInfo.Status status);
```
I see. I didn't realize that we had added the setter methods to the `ManifestEntry` interface. Looks like that was probably allowed because `ManifestEntry` is in core and not exposed in the public API. Instead, `DataFile` and `DeleteFile` are in the API.

Here, I think we need to be more careful. This is very likely going to be in API along side `DataFile`, so it should be an immutable interface. To avoid downcasting, the manifest reader should configure its Parquet reader to produce the concrete class, `GenericTrackedFile`. That class can be mutable so the reader can pass instances to `InheritableTrackedMetadata` and then the reader should return those instances as `TrackedFile` (because it is `CloseableIterable<TrackedFile>`). That shouldn't require casting.

rdblue commented on api/src/main/java/org/apache/iceberg/TrackingInfo.java:
```diff
@@ -0,0 +1,118 @@
+ */
+public interface TrackingInfo {
+  /** Status of an entry in a tracked file */
+  enum Status {
```
Okay, this is fine for now. We may want to move it later, but we'll see what makes sense.

rdblue commented on api/src/main/java/org/apache/iceberg/TrackingInfo.java:
```diff
@@ -0,0 +1,118 @@
+          1,
+          "snapshot_id",
+          Types.LongType.get(),
+          "Snapshot ID where the file was added, or deleted if status is 2. Inherited when null.");
```
I think this should be `"Snapshot ID where the file was added or deleted/removed"`. No need to mention specific status codes or inheritance. People reading these docs should never see values without a snapshot ID because of inheritance so I don't think there is value in adding it here because this is primarily for metadata tables.

rdblue commented on api/src/main/java/org/apache/iceberg/TrackingInfo.java:
```diff
@@ -0,0 +1,118 @@
+   *
+   * <p>If content_type is 0 (DATA), this is the _row_id for the first row in the data file. If
+   * content_type is 3 (DATA_MANIFEST), this is the starting _row_id to assign to rows added by
+   * ADDED data files.
```
As with the schema fields, I don't think that it is valuable to describe the format behavior here, either.

rdblue commented on api/src/main/java/org/apache/iceberg/TrackedFile.java:
```diff
@@ -0,0 +1,280 @@
+  void setFirstRowId(Long firstRowId);
+
+  /** Set the ordinal position in the manifest. */
+  void setPos(Long position);
```
I don't think that this is needed. In the v3 implementation, we use a column that projects _pos without calling a setter.

rdblue commented on core/src/main/java/org/apache/iceberg/InheritableTrackedMetadataFactory.java:
```diff
@@ -0,0 +1,94 @@
+   * @param manifestEntry the DATA_MANIFEST or DELETE_MANIFEST tracked file from root
+   * @return inheritable metadata instance
+   */
+  static InheritableTrackedMetadata fromTrackedFile(TrackedFile<?> manifestEntry) {
```
Methods like this should accept a typed file, like `ManifestFile`, instead of converting. That way we can keep the conversion logic in a single place and the caller can either pass an existing `ManifestFile` or calls `trackedFile.asManifestFile()`.

anoopj replied:
Removed it for now.

rdblue commented on core/src/main/java/org/apache/iceberg/InheritableTrackedMetadataFactory.java:
```diff
@@ -0,0 +1,94 @@
+    Long sequenceNumber = tracking != null ? tracking.sequenceNumber() : null;
+
+    Preconditions.checkArgument(
+        snapshotId != null, "Manifest entry must have snapshot ID: %s", manifestEntry.location());
```
The term "manifest entry" is confusing in v4 because it could be referring to an entry in a manifest (the original definition) or a tracked file for a manifest. This is why we are using the term "tracked file" now. Method names, variable names, and error messages for v4 metadata should always replace "manifest entry" with a more specific and less ambiguous term.

rdblue commented on core/src/main/java/org/apache/iceberg/TrackedDataFileAdapter.java:
```diff
@@ -0,0 +1,186 @@
+  private final PartitionSpec spec;
+
+  TrackedDataFileAdapter(TrackedFile<?> trackedFile, PartitionSpec spec) {
+    if (trackedFile.contentType() != FileContent.DATA) {
```
I would rather have this check in `asDataFile` to avoid allocation before running the check. I'd also probably make this a private inner class of `TrackedFile` to avoid exposing it and needing to lock down the constructor and check the content like this.

rdblue commented on core/src/main/java/org/apache/iceberg/TrackedDataFileAdapter.java:
```diff
@@ -0,0 +1,186 @@
+
+  @Override
+  public String manifestLocation() {
+    return trackedFile.manifestLocation();
```
Using an inner class would allow you to reference the private variables directly.

rdblue commented on core/src/main/java/org/apache/iceberg/TrackedDataFileAdapter.java:
```diff
@@ -0,0 +1,186 @@
+
+  @Override
+  public StructLike partition() {
+    return null;
```
This should be marked with a TODO. Implementing this on top of content stats is a key part of validating the v4 partition design.

rdblue commented on core/src/main/java/org/apache/iceberg/TrackedDataFileAdapter.java:
```diff
@@ -0,0 +1,186 @@
+  @Override
+  public long fileSizeInBytes() {
+    Long size = trackedFile.fileSizeInBytes();
+    return size != null ? size : 0L;
```
Iceberg should never lie about metadata. Instead, allow this to throw a null pointer exception.

rdblue commented on core/src/main/java/org/apache/iceberg/TrackedDataFileAdapter.java:
```diff
@@ -0,0 +1,186 @@
+
+  @Override
+  public Map<Integer, Long> columnSizes() {
+    return null;
```
Also TODO. It's important to adapt from the new metadata structure to the old.

rdblue commented on core/src/main/java/org/apache/iceberg/TrackedDataFileAdapter.java:
```diff
@@ -0,0 +1,186 @@
+  @Override
+  public DataFile copyWithStats(Set<Integer> requestedColumnIds) {
+    return new TrackedDataFileAdapter(
+        (TrackedFile<?>) trackedFile.copyWithStats(requestedColumnIds), spec);
```
Why is a cast necessary here? This is suspicious; probably caused by the type param on `TrackedFile`. Hopefully we can remove this when we fix that issue.

anoopj replied:
It was unrelated to the type param. Fixed it with a refactoring.

rdblue commented on core/src/main/java/org/apache/iceberg/GenericTrackedFile.java:
```diff
@@ -0,0 +1,773 @@
+    implements TrackedFile<GenericTrackedFile>,
+        IndexedRecord,
+        StructLike,
+        SpecificData.SchemaConstructable,
```
There's no need for Avro interfaces since we no longer use Avro generics to read metadata. You can remove `IndexedRecord` because it is replaced by `StructLike` and remove `SpecificData.SchemaConstructable` because that's handled by [reflection in the readers](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/avro/InternalReaders.java#L56-L59).

rdblue commented on core/src/main/java/org/apache/iceberg/GenericTrackedFile.java:
```diff
@@ -0,0 +1,773 @@
+          ManifestStats.ADDED_ROWS_COUNT.asOptional(),
+          ManifestStats.EXISTING_ROWS_COUNT.asOptional(),
+          ManifestStats.DELETED_ROWS_COUNT.asOptional(),
+          ManifestStats.MIN_SEQUENCE_NUMBER.asOptional());
```
Each of the sub-structs should be its own struct type (like `ManifestStatsStruct` and `TrackingInfoStruct`) that is read and written the same way that this class is. The `TrackedFile` schema should embed each sub struct as a field.

rdblue commented on core/src/main/java/org/apache/iceberg/GenericTrackedFile.java:
```diff
@@ -0,0 +1,773 @@
+   * @param toCopy a tracked file to copy
+   * @param copyStats whether to copy stats
+   */
+  private GenericTrackedFile(GenericTrackedFile toCopy, boolean copyStats) {
```
For v4, we can remove the paths that use a `copyStats` boolean. Instead, we should use a set of field IDs to identify the fields to copy, much like the newer copy API.

rdblue commented on core/src/main/java/org/apache/iceberg/GenericTrackedFile.java:
```diff
@@ -0,0 +1,773 @@
+  private String referencedFile = null;
+
+  // Manifest stats (for manifest entries)
+  private Integer addedFilesCount = null;
```
These fields are required in v4, so they should be primitives in `ManifestStatsStruct`.

rdblue commented on core/src/main/java/org/apache/iceberg/GenericTrackedFile.java:
```diff
@@ -0,0 +1,773 @@
+
+  // Deletion vector fields
+  private Long deletionVectorOffset = null;
+  private Long deletionVectorSizeInBytes = null;
```
These should be required in the DV struct, since the struct itself is optional.

rdblue commented on core/src/main/java/org/apache/iceberg/V4ManifestReader.java:
```diff
@@ -0,0 +1,219 @@
+ * represent data files, delete files, or manifest references. TODO: implement caching.
+ */
+class V4ManifestReader extends CloseableGroup implements CloseableIterable<TrackedFile<?>> {
+  static final ImmutableList<String> ALL_COLUMNS = ImmutableList.of("*");
```
The original `ManifestReader` allows callers to project columns by calling `select(String... columns)`, but this isn't a very useful API because the majority of the time that a manifest is being projected, it is by the Iceberg library for a specific purpose. That's why the arguments to `select` are usually `"*"`, `BaseScan.SCAN_WITH_STATS_COLUMNS`, or `BaseScan.SCAN_COLUMNS`. And now that we need to project columnar stats based on fields used in filters, the last two constants are no longer useful.

I think that the v4 implementation should provide more useful projection capabilities:
1. By default, all columns are projected, but without needing to check for `*`
2. A filter-based stats projection that is automatic, based on the filters pushed down to this reader
3. A custom stats projection like `projectStats(Set<Integer> fieldIds)` that sets the content stats projection (to be unioned with the fields from #2)
4. A `project(Schema)` method that can perform more fine-grained projection when the reader is used in metadata table scans and other custom scan situations (like `ManifestFiles.readPaths`).

Removing `select` from the API allows us to simplify the logic in this reader. We don't need to keep lists of fields to pass in or detect when both `select` and `project` are called.

anoopj replied:
That sounds good.

rdblue commented on core/src/main/java/org/apache/iceberg/V4ManifestReader.java:
```diff
@@ -0,0 +1,219 @@
+    return this;
+  }
+
+  public CloseableIterable<TrackedFile<?>> entries() {
```
In v4, we are inverting the relationship between an entry and a file. Now, a tracked file contains its "entry" metadata as tracking info. The purpose of that change is to avoid needing additional methods to get the "entries" versus getting the "files".

The reader can now be simplified. It needs two read capabilities: `allFiles` and `liveFiles` (which filters out removed/deleted files). The `ClosableIterable` interface should use `liveFiles` because that's the simplest generic contract: when a manifest is read you don't return deleted files unless they are requested.

rdblue commented on core/src/main/java/org/apache/iceberg/V4ManifestReader.java:
```diff
@@ -0,0 +1,219 @@
+  private CloseableIterable<TrackedFile<?>> open(Collection<String> cols) {
+    Schema projection = buildProjection(cols);
+
+    FileFormat format = FileFormat.fromFileName(file.location());
```
This happens in the old reader because there was no reason to plumb the format through. In v4, this should be passed into the manifest reader instead.

rdblue commented on core/src/main/java/org/apache/iceberg/V4ManifestReader.java:
```diff
@@ -0,0 +1,219 @@
+    try {
+      bitmap.deserialize(new DataInputStream(new ByteArrayInputStream(bytes)));
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to deserialize Roaring bitmap from manifest DV");
```
Prefer the built-in `UncheckedIOException` to `RuntimeIOException`.

rdblue commented on core/src/main/java/org/apache/iceberg/V4ManifestReader.java:
```diff
@@ -0,0 +1,219 @@
+
+    ByteBuffer buffer = dvInfo.inlineContent();
+    byte[] bytes = new byte[buffer.remaining()];
+    buffer.asReadOnlyBuffer().get(bytes);
```
Use `ByteBuffers.toByteArray` instead of copying directly.

rdblue commented on core/src/main/java/org/apache/iceberg/V4ManifestReader.java:
```diff
@@ -0,0 +1,219 @@
+    addCloseable(entries);
+
+    CloseableIterable<TrackedFile<?>> transformed =
+        CloseableIterable.transform(entries, inheritableMetadata::apply);
```
Actually, this is not true. Row ID assignment needs to happen first, then rows can be filtered.

anoopj committed:
Incorporate PR feedback from Ryan

anoopj commented on core/src/main/java/org/apache/iceberg/GenericTrackedFile.java:
```diff
@@ -0,0 +1,773 @@
+import org.apache.iceberg.util.ByteBuffers;
+
+/** Generic implementation of {@link TrackedFile} for V4 manifests. */
+public class GenericTrackedFile extends SupportsIndexProjection
```
Done.

anoopj commented on core/src/main/java/org/apache/iceberg/InheritableTrackedMetadata.java:
```diff
@@ -0,0 +1,36 @@
+   * @param entry the tracked file entry
+   * @return the entry with metadata applied
+   */
+  TrackedFile<?> apply(TrackedFile<?> entry);
```
Thanks. I've implemented something along the lines of your suggestion. PTAL.

anoopj committed:
Incorporate feedback

anoopj committed:
Improve V4ManifestReader based on PR review feedback

  - Replace select() with filterData(), projectStats(), project() methods
  - Remove firstRowId from readRoot(); only leaf manifests need assignment
  - Use MetadataColumns.ROW_POSITION for safe position tracking
  - Always project tracking_info for correctness
  - Add WRITE_TYPE schema excluding read-only metadata columns

anoopj committed:
Change asDataFile/asDeleteFile to no-argument API

anoopj committed:
Update TrackedFile to use ContentStats

- Change TrackedFile.contentStats() return type from Object to ContentStats
- Add helper methods to convert ContentStats to Map-based stats for
  DataFile/DeleteFile compatibility
- TrackedDataFile and TrackedDeleteFile now properly translate:
  - valueCounts
  - nullValueCounts
  - nanValueCounts
  - lowerBounds
  - upperBounds
- Add filterContentStats for copyWithStats support

anoopj commented:
still active
